### PR TITLE
Migrating the rendering openGL calls from DeferredLightingEffect to gpu::Batch

### PR DESCRIPTION
**Pull request by samcake**
- Migrate the rendering of the deferredLightingEffect from naked GL to using gpu library

- Add the true suport for Framebuffer and clears in gpu

- Fix a bug with spot lights and the volume geometry used which was failing for spot angles of 90degrees
 We provide a combination of normalized cone geometry and vertex shader that adjust for all angles

- Provide the starting class for a library of commonly used shaders in gpu::StandardShaderLib

- Provide shortcut const names for very frequently used Format::Element

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6b229d8d2d785067b031cdc3a3d5891bdab0a013 available at: https://gist.github.com/e2b03ee6df2ad00dc598<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.dmg

File: libraries/render-utils/src/DeferredLightingEffect.cpp [hunk 1]
```diff
@@ -393,148 +400,168 @@ void DeferredLightingEffect::render(RenderArgs* args) {
...
-                
-                glMatrixMode(GL_PROJECTION);
-                glPushMatrix();
-                glLoadIdentity();
-                
-                renderFullscreenQuad();
-            
-                glPopMatrix();
-                glMatrixMode(GL_MODELVIEW);
+            // TODO: We shouldn;t have to do that test and use a different volume geometry for when inside the vlight volume,
```
Comment (ZappoMan) [inline libraries/render-utils/src/DeferredLightingEffect.cpp]: dead code? should we delete this?

File: libraries/render-utils/src/DeferredLightingEffect.cpp [hunk 2]
```diff
@@ -393,148 +400,168 @@ void DeferredLightingEffect::render(RenderArgs* args) {
...
+        for (auto lightID : _spotLights) {
+            auto light = _allocatedLights[lightID];
+            // IN DEBUG:
+             light->setShowContour(true);
+            batch.setUniformBuffer(_spotLightLocations.lightBufferUnit, light->getSchemaBuffer());
             
+            auto eyeLightPos = eyePoint - light->getPosition();
+            auto eyeHalfPlaneDistance = glm::dot(eyeLightPos, light->getDirection());
+            
+            glm::vec4 coneParam(light->getSpotAngleCosSin(), 0.66f * tan(0.5 * light->getSpotAngle()), 1.0f);
```
Comment (ZappoMan) [inline libraries/render-utils/src/DeferredLightingEffect.cpp]: magic numbers? why 0.66?

File: libraries/render-utils/src/DeferredLightingEffect.h [hunk 3]
```diff
@@ -94,15 +95,49 @@ class DeferredLightingEffect : public Dependency {
...
+    gpu::PipelinePointer _directionalLightShadowMap;
+    LightLocations _directionalLightShadowMapLocations;
+    gpu::PipelinePointer _directionalLightCascadedShadowMap;
+    LightLocations _directionalLightCascadedShadowMapLocations;
+
+    gpu::PipelinePointer _pointLight;
+    LightLocations _pointLightLocations;
+    gpu::PipelinePointer _spotLight;
+    LightLocations _spotLightLocations;
+/*
```
Comment (ZappoMan) [inline libraries/render-utils/src/DeferredLightingEffect.h]: dead code?

File: libraries/render-utils/src/deferred_light_limited.slv [hunk 4]
```diff
@@ -12,9 +12,23 @@
...
 
+<@include gpu/Transform.slh@>
+
+<$declareStandardTransform()$>
+
+uniform mat4 texcoordMat;
+
 void main(void) {
-    gl_Position = ftransform();
+  //  gl_Position = ftransform();
```
Comment (ZappoMan) [inline libraries/render-utils/src/deferred_light_limited.slv]: dead code?

File: libraries/render-utils/src/deferred_light_limited.slv [hunk 5]
```diff
@@ -12,9 +12,23 @@
...
+
+    // standard transform
+    TransformCamera cam = getTransformCamera();
+    TransformObject obj = getTransformObject();
+    <$transformModelToClipPos(cam, obj, gl_Vertex, gl_Position)$>;
+
     vec4 projected = gl_Position / gl_Position.w;
-    gl_TexCoord[0] = vec4(dot(projected, gl_ObjectPlaneS[3]) * gl_Position.w,
-        dot(projected, gl_ObjectPlaneT[3]) * gl_Position.w, 0.0, gl_Position.w);
+   // gl_TexCoord[0] = vec4(dot(projected, gl_ObjectPlaneS[3]) * gl_Position.w,
```
Comment (ZappoMan) [inline libraries/render-utils/src/deferred_light_limited.slv]: dead code?

File: libraries/render-utils/src/spot_light.slf [hunk 6]
```diff
@@ -28,13 +28,16 @@ void main(void) {
...
     DeferredFragment frag = unpackDeferredFragment(texCoord);
-      
+
     // Kill if in front of the light volume
     float depth = frag.depthVal;
     if (depth < gl_FragCoord.z) {
         discard;
     }
 
+    //gl_FragColor = vec4(1.0, 1.0, 1.0, 0.0);
```
Comment (ZappoMan) [inline libraries/render-utils/src/spot_light.slf]: dead code?

**Comment by hifi-gustavo** [comment]
Build console output for commit: e901d5cad0c5bfdc31a0a14e160d99577d2e0bfe available at: https://gist.github.com/e2b03ee6df2ad00dc598<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: fb37af3e043a94e091a41742f384745d9493bc64 available at: https://gist.github.com/e2b03ee6df2ad00dc598<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.dmg

**Comment by ZappoMan** [comment]
How should I test this? Do you have some test content?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8f4a10844195120cdaa19f9675f7fc9dac30a30d available at: https://gist.github.com/e2b03ee6df2ad00dc598<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.dmg

File: libraries/render-utils/src/DeferredLightingEffect.cpp (hunk 2 above)
Comment (samcake) [inline libraries/render-utils/src/DeferredLightingEffect.cpp]: 2/3 daemon ?

**Comment by samcake** [comment]
You can go to my domain "asimov" next to me and could validate that the 2 lights cone and point lights are working correctly
in particular for spot lights, if the angle is 90 deg then it should look like an half sphere lighting of course

Or you can just create a point light and a spot light and check that it's working correctly

File: libraries/gpu/src/gpu/Format.h [hunk 7]
```diff
@@ -203,6 +206,19 @@ enum ComparisonFunction {
...
 };
 
+enum Primitive {
+    POINTS = 0,
+    LINES,
+    LINE_STRIP,
+    TRIANGLES,
+    TRIANGLE_STRIP,
+    TRIANGLE_FAN,
+    QUADS,
```
Comment (jherico) [inline libraries/gpu/src/gpu/Format.h]: we'll need to drop quads and quad strips before moving to core profile.

Comment (samcake) [inline libraries/gpu/src/gpu/Format.h]: Wordss of wisdom !
You hear that @ZappoMan !

**Comment by hifi-gustavo** [comment]
Build console output for commit: 110b009aa4fc312c89cad93015fcd953ca303417 available at: https://gist.github.com/e2b03ee6df2ad00dc598<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.dmg

File: libraries/gpu/src/gpu/Format.h (hunk 7 above)
Comment (ZappoMan) [inline libraries/gpu/src/gpu/Format.h]: GAK!

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2968286779d72b67c1c01c27fe87a24af7fcf472 available at: https://gist.github.com/e2b03ee6df2ad00dc598<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.dmg

**Comment by samcake** [comment]
Play again, fixed the issue on Mac

**Comment by ZappoMan** [comment]
I'll test shortly - once build is complete!

**Comment by samcake** [comment]
tested on windows too successfully

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1f13106f5167b947c177e45be6c3d97450c575c6 available at: https://gist.github.com/e2b03ee6df2ad00dc598<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5286.dmg

**Comment by ZappoMan** [comment]
testing now

**Pull request merged**